resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,575 @@
+					bulkhead.onSuccess(start, TimeUnit.MILLISECONDS);
+				}
+			}
+		};
+	}
+
+
+
+	/**
+	 * Create a Bulkhead with a default configuration.
+	 *
+	 * @param name the name of the bulkhead
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead ofDefaults(String name) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, AdaptiveBulkheadConfig.ofDefaults());
+	}
+
+
+	/**
+	 * Creates a bulkhead with a custom configuration and custom limiter
+	 *
+	 * @param name         the name of the bulkhead
+	 * @param config       a custom BulkheadConfig configuration
+	 * @param limitAdapter the custom limit adopter
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config, LimitPolicy limitAdapter) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config, limitAdapter);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name   the name of the bulkhead
+	 * @param config a custom BulkheadConfig configuration
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name                   the name of the bulkhead
+	 * @param bulkheadConfigSupplier custom configuration supplier
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, Supplier<AdaptiveBulkheadConfig> bulkheadConfigSupplier) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, bulkheadConfigSupplier.get());
+	}
+
+	interface AdaptiveMetrics extends Bulkhead.Metrics {
+		/**
+		 * Returns the current total number of calls which were slower than a certain threshold.
+		 *
+		 * @return the current total number of calls which were slower than a certain threshold
+		 */
+		int getNumberOfSlowCalls();
+
+		/**
+		 * Returns the current number of failed buffered calls in the ring buffer.
+		 *
+		 * @return the current number of failed buffered calls in the ring buffer
+		 */
+		int getNumberOfFailedCalls();
+
+		/**
+		 * Returns the current number of successful buffered calls in the ring buffer.
+		 *
+		 * @return the current number of successful buffered calls in the ring buffer
+		 */
+		int getNumberOfSuccessfulCalls();
+
+		/**
+		 * @return average latency for service calls in millis
+		 */
+		double getAverageLatencyMillis();
+
+		/**
+		 * @return current failure rate for recorded calls if error check is enabled for related exceptions
+		 */
+		double getFailureRate();
+
+		/**
+		 * @return current slow call rate for the recorded calls
+		 */
+		double getSlowCallRate();
+	}
+
+	/**
+	 * An EventPublisher which can be used to register event consumers.
+	 */
+	interface AdaptiveEventPublisher extends io.github.resilience4j.core.EventPublisher<BulkheadLimit> {
```

Comment:
Can we rename the base interface to `AdaptiveBulkeadEvent` comparable to `CircuitBreakerEvent` or `BulkheadEvent`